Crash when allocating unreasonable amounts of texture memory

nical opened the pull request:
Track texture allocations and crash if we go beyond a certain budget. The budget (2Gb by default) is configurable in the renderer options. I'll experiment with various thresholds in gecko (which probably need to take into account the window size) when this hits mozilla-central.

Fixes #3207.

gw3583 commented:
Looks like some of the PF code needs updating:

```
error[E0308]: mismatched types
  --> webrender\src\gpu_glyph_renderer.rs:66:41
   |
66 |         device.upload_texture_immediate(&area_lut_texture, area_lut_pixels);
   |                                         ^^^^^^^^^^^^^^^^^ expected struct `device::gl::Texture`, found enum `std::result::Result`
   |
   = note: expected type `&device::gl::Texture`
              found type `&std::result::Result<device::gl::Texture, device::gl::AllocError>`
error[E0308]: mismatched types
  --> webrender\src\gpu_glyph_renderer.rs:90:13
   |
90 |             area_lut_texture,
   |             ^^^^^^^^^^^^^^^^ expected struct `device::gl::Texture`, found enum `std::result::Result`
   |
   = note: expected type `device::gl::Texture`
              found type `std::result::Result<device::gl::Texture, device::gl::AllocError>`
error[E0308]: mismatched types
   --> webrender\src\gpu_glyph_renderer.rs:129:13
    |
129 |             texture,
    |             ^^^^^^^ expected struct `device::gl::Texture`, found enum `std::result::Result`
    |
    = note: expected type `device::gl::Texture`
               found type `std::result::Result<device::gl::Texture, device::gl::AllocError>`
error[E0308]: mismatched types
   --> webrender\src\gpu_glyph_renderer.rs:191:46
    |
191 |         self.device.upload_texture_immediate(&path_info_texture, &path_info_texels);
    |                                              ^^^^^^^^^^^^^^^^^^ expected struct `device::gl::Texture`, found enum `std::result::Result`
    |
    = note: expected type `&device::gl::Texture`
               found type `&std::result::Result<device::gl::Texture, device::gl::AllocError>`
error[E0308]: mismatched types
   --> webrender\src\gpu_glyph_renderer.rs:229:60
    |
229 |         self.device.bind_texture(TextureSampler::color(1), &path_info_texture);
    |                                                            ^^^^^^^^^^^^^^^^^^ expected struct `device::gl::Texture`, found enum `std::result::Result`
    |
    = note: expected type `&device::gl::Texture`
               found type `&std::result::Result<device::gl::Texture, device::gl::AllocError>`
error[E0308]: mismatched types
   --> webrender\src\gpu_glyph_renderer.rs:234:36
    |
234 |         self.device.delete_texture(path_info_texture);
    |                                    ^^^^^^^^^^^^^^^^^ expected struct `device::gl::Texture`, found enum `std::result::Result`
    |
    = note: expected type `device::gl::Texture`
               found type `std::result::Result<device::gl::Texture, device::gl::AllocError>`
error: aborting due to 6 previous errors
For more information about this error, try `rustc --explain E0308`.

```

nical commented:
I fixed the failing rawtest and added a commit that prints a message and skips the rawtests that rely on a the device pixel ratio being 1.0 when it is not the case to avoid having to comment them out locally.

nical commented:
Try push: https://treeherder.mozilla.org/#/jobs?repo=try&revision=965f427b69641a989091c8f5eedb2c981e061559

gw3583 commented:
Generally looks good to me - should we change the ```unwrap``` calls to ```expect``` to provide a bit more context on any crashes that do occur?

nical commented on webrender/src/device/gl.rs:
```diff
@@ -1583,6 +1609,8 @@ impl Device {
             self.release_depth_target(texture.get_dimensions());
         }
 
+        self.deallocated_gpu_memory(texture.size_in_bytes());
```
I didn't find any. Our code to resize texture currently destroys and re-creates bigger textures (with possibly a blit from the old texture to the new one.

bholley replied:
Yeah, this was a change I had to make for immutable storage. Texture re-initialization is no more.

nical commented on webrender/src/renderer.rs:
```diff
@@ -5043,6 +5059,47 @@ impl Renderer {
+        }
+    }
+
+    fn log_allocation_failure(&self, error: AllocError) {
```
Since the budget is there to force a crash when things are so wrong that it's best not to keep the browser alive, I don't think it is very interesting to display.
That said I think that displaying some number that we think is reasonable for an average page, or some number beyond which we believe something worth investigating (but not to the point that we'd crash the browser) would be useful, as well as displaying the current allocated texture memory.

kvark commented:
That try push needs another try

nical commented:
> That try push needs another try

Ah, yes indeed, there we go: https://treeherder.mozilla.org/#/jobs?repo=try&revision=56a80fb4e9602775ab0530099c9c443ba43b26fa

bholley commented:
Per discussion on IRC late last week, we should rebase it on top of [1] when that lands to avoid having multiple mechanisms for tracking GPU memory usage.

[1] https://bugzilla.mozilla.org/show_bug.cgi?id=1504115

bors-servo commented:
:umbrella: The latest upstream changes (presumably #3269) made this pull request unmergeable. Please resolve the merge conflicts.

gw3583 commented:
Looks like the bug mentioned above has landed - so I guess this needs a rebase on top of that?

gw3583 commented:
@nical @bholley Do we just need to get this rebased and then it's ready to land?

bholley commented:
It just needs to be rebased to use the global atomic byte count rather than tracking it per-device. That should be trivial, then it can land.

gw3583 commented:
ping @nical - should we get this rebased and landed?
